src/base/static/libs/maps/mapboxgl-provider.js

```diff
@@ -29,10 +152,62 @@ export default (container, options) => {
       map.on(event, callback);
     },
 
+    off: (event, callback) => {
+      map.off(event, callback);
+    },
+
+    onLayerEvent: (event, layerId, callback, context) => {
+      callback = callback.bind(context);
+      map.on(event, layerId, callback);
+    },
+
+    bindPlaceLayerEvent: (eventName, layer, callback) => {
+      ["symbol", "fill", "line"].forEach(layerGeometryType => {
+        const layerId = `${layer.id}_${layerGeometryType}`;
+        let topmostProperties;
+        map.on(eventName, layerId, evt => {
+          if (eventName === "click") {
+            // For click events, we query rendered features here to obtain a
+            // single array of layers below the clicked-on point. The first
+            // entry in this array corresponds to the topmost rendered feature.
+            topmostProperties = map.queryRenderedFeatures(
+              [evt.point.x, evt.point.y],
+              {
+                // Limit these click listeners to place geometry.
+                filter: ["==", ["get", "type"], "place"],
+              },
+            )[0].properties;
+          }
+
+          callback(topmostProperties);
```

Comment:
Is `.properties` here referencing the GeoJSON properties? Perhaps we shouldn't return `.properties`, but the entire layer. Then the caller can decide what to do with it, and we wouldn't have to keep track of which parts of the layer we are passing around, but instead treat the layer as first class.

Reply:
I agree-- we can just return the layer itself in this case.